django/db/models/fields/__init__.py

```diff
@@ -135,12 +135,15 @@ def __init__(self, verbose_name=None, name=None, primary_key=False,
                  db_index=False, rel=None, default=NOT_PROVIDED, editable=True,
                  serialize=True, unique_for_date=None, unique_for_month=None,
                  unique_for_year=None, choices=None, help_text='', db_column=None,
-                 db_tablespace=None, auto_created=False, validators=(),
+                 db_tablespace=None, auto_created=False, returning=False, validators=(),
                  error_messages=None):
         self.name = name
         self.verbose_name = verbose_name  # May be set by set_attributes_from_name
         self._verbose_name = verbose_name  # Store original for deconstruction
         self.primary_key = primary_key
+        # `returning` is a private API, subject to change. Currently only
+        # PostgreSQL supports returning multiple fields on a model.
+        self.returning = returning or (primary_key and default is NOT_PROVIDED)
```

Comment:
I'd vote for making `returning` a `property` instead of a stealth field option at least for now because this is not something we've done in the past.

```python
@property
def returning(self):
    return hasattr(self.default, 'as_sql')

AutoField.returning = True
```

That would make `DateTimeField(default=Now)` work and avoid the ambiguity of `default=Now, returning=False`. We'd still have to deal with backends that don't support returning fields.

Reply:
Sorry, I don't think I fully understand your reasoning. What is the difference between a property or an instance attribute? Isn't the only difference, that the former can't be set?

So what I think you want to achieve, is that you can't set `default=Now` and set returning to False. Right? I think we are getting ahead of ourselves here. This patch does not intent to support database defaults, at least not as a public API.

Anyhow, here is my suggestion. I will remove the `returning` kwarg from the constructor. That way you can't set it at all. It will remain something only exposed to custom fields. Sounds good?

Reply:
> So what I think you want to achieve, is that you can't set default=Now and set returning to False. Right?

I'm not saying I want to achieve that. I'm saying that the API you are suggesting, even if private, allows undefined and untested behaviour when `default=Now, returning=False` is set.

> I think we are getting ahead of ourselves here. This patch does not intent to support database defaults, at least not as a public API.

@codingjoe my main concern with adding a model field kwargs is the probability we'll have to deprecate it in the future in favor of something that is more though of and in line with the previous discussions (https://github.com/django/deps/pull/36, #5904, #2149). For example the ORM usually tries to abstract implementation details and prefix database related options with `db_*`.

Adding a stealth model field kwarg or deprecating one is not something we've done in the past and I assume it's going to be a lot more pain than deprecating a field attribute. We still have to keep shims around for old deprecated fields to avoid breaking migrations for example.

> Anyhow, here is my suggestion. I will remove the returning kwarg from the constructor. That way you can't set it at all. It will remain something only exposed to custom fields. Sounds good?

If we can avoid using a kwarg for this I think it's a safe way forward. I think it'd even be better if all the returning logic would be pushed down to the compiler without defining any model field attributes for now. Having the compiler automatically build it's `RETURNING` clause based on expressions that are db generated would be even better IMO.

Reply:
> @codingjoe my main concern with adding a model field kwargs is the probability we'll have to deprecate it in the future in favor of something that is more though of and in line with the previous discussions (django/deps#36, #5904, #2149). For example the ORM usually tries to abstract implementation details and prefix database related options with db_*.

Good point, I will rename the attributes to `db_returning`. That does seem be a more descriptive name.

> If we can avoid using a kwarg for this I think it's a safe way forward. I think it'd even be better if all the returning logic would be pushed down to the compiler without defining any model field attributes for now. Having the compiler automatically build it's RETURNING clause based on expressions that are db generated would be even better IMO.

I believe the change I made actually goes into that direction, even if not fully meeting that requirement. See, previously the compiler did know what a model has to have primary key and made sure to return it. Now the compiler is just being told, "hey, insert that! oh and while you are at it, I what is the value of the `id` field?"
The `RETURNING` statement is compiled by the compiler, the only thing that the compiler does not know, is which fields should be returned and which not. I know if would be super elegant to go with the `.as_sql` approach for default values. As cool as that might be, I believe it should be done in a separate commit and preferably after gaining some experience with some actual implementations of this feature.